pages/architecture/session_tokens_rfc.rst

```diff
@@ -0,0 +1,152 @@
+.. _nuts-documentation-session-tokens:
+
+RFC: Session Tokens
+###################
+
+Session tokens are tokens which should be provided along an API request which refer
+to contextual information of the request.
+A session is scoped to an actor, subject and custodian. Optionally a reference
+to a specific consent record can be provided.
+
+Motivation
+**********
+
+Originally the plan was to only provide the Actor's
+:ref:`nuts-documentation-login-contracts` along the API request, assuming this
+was sufficient. However we encountered situations where more information was
+needed.
+
+Size
+####
+
+The more attributes are requested to sign the IMRA login contract, the bigger
+the resulting JSON hash gets. So big that when base64 encoded and signed as a
+JWTit exceeds the often used http header size of 8kb. Although header sizes are
+not officially limited, approaching this limit will likely result in unforeseen
+problems.
+
+Subject scoping
+###############
+
+How does the custodians server knows about which subject the data request is
+about? Some protocols contain a patient id in the URL, others require inspecting
+the body or headers. In order for Nuts to have a minimal implementation effort,
+is is important to refer to the patient in a uniform way.
+
+Custodian scoping
+#################
+
+Some protocols run on a multi tenant environment which need a way of specifying
+the custodian the request is about. Just like the subject, in order for Nuts to
+have a minimal implementation effort, is is important to refer to the custodian
+in a uniform way.
+
+Cost of checking consent and Identity
+#####################################
+
+When performing multiple request for a triple of custodian|subject|actor, the
+:ref:`nuts-documentation-login-contracts`, and consent must be verified each
+time, increasing response time.
+
+Limitations
+###########
+
+Because it is necessary to request a new session for every combination of subject
+custodian and actor, bulk operations, like getting reports for many patients
+at once, is not possible. We could consider allowing list of subjects during the
+token requests.
+
+Mechanics
+*********
+
+To resolve above issues, we'll introduce sessions. A session lives on the
+custodian side and contains information about the subject, actor and custodian.
+Additional information can be provided like, session expiration and consent reference.
+
+To refer to a session during an API call, a reference to the session must be provided.
+For example, in the case of a HTTP REST call in the form of a bearer token in the
+authorization header.
+
+The session token can be provided using a OAuth 2.0 flow with the
+`urn:ietf:params:oauth:grant-type:jwt-bearer` grant_type. This profile is described in `RFC7523 <https://tools.ietf.org/html/rfc7523>`_.
+
+The returned session token can be a reference to a session or a signed JWT
+containing all the relevant information needed by the API server.
```

Comment:
Maybe the role 'API server' should be referred to from the diagram, so that it is clear who's who?